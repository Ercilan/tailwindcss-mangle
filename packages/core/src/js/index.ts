import type { StringLiteral, TemplateElement } from '@babel/types'
import MagicString from 'magic-string'
import { jsStringEscape } from '@ast-core/escape'
import type { IJsHandlerOptions } from '../types'
import { makeRegex, splitCode } from '../shared'
import { parse, traverse } from '@/babel'

export { preProcessJs, preProcessRawCode } from './pre'

export function handleValue(raw: string, node: StringLiteral | TemplateElement, options: IJsHandlerOptions, ms: MagicString, offset: number, escape: boolean) {
  const { ctx, splitQuote = true } = options
  const { replaceMap, classGenerator: clsGen } = ctx

  const array = splitCode(raw, {
    splitQuote,
  })
  let rawString = raw
  let needUpdate = false
  for (const v of array) {
    if (replaceMap.has(v)) {
      let ignoreFlag = false
      if (Array.isArray(node.leadingComments)) {
        ignoreFlag = node.leadingComments.findIndex(x => x.value.includes('tw-mangle') && x.value.includes('ignore')) > -1
      }

      if (!ignoreFlag) {
        rawString = rawString.replace(makeRegex(v), clsGen.generateClassName(v).name)
        needUpdate = true
      }
    }
  }
  if (needUpdate && typeof node.start === 'number' && typeof node.end === 'number') {
    const start = node.start + offset
    const end = node.end - offset

    if (start < end && raw !== rawString) {
      ms.update(start, end, escape ? jsStringEscape(rawString) : rawString)
    }
  }
  return rawString
}

export function jsHandler(rawSource: string | MagicString, options: IJsHandlerOptions) {
  const ms: MagicString = typeof rawSource === 'string' ? new MagicString(rawSource) : rawSource
  const ast = parse(ms.original, {
    sourceType: 'unambiguous',
    plugins: ['typescript', 'jsx'],
  })
  traverse(ast, {
    StringLiteral: {
      enter(p) {
        const n = p.node
        handleValue(n.value, n, options, ms, 1, true)
      },
    },
    TemplateElement: {
      enter(p) {
        const n = p.node
        handleValue(n.value.raw, n, options, ms, 0, false)
      },
    },
    // CallExpression: {
    //   enter(p: NodePath<CallExpression>) {
    //     const calleePath = p.get('callee')
    //     if (calleePath.isIdentifier() && calleePath.node.name === 'eval') {
    //       p.traverse({
    //         StringLiteral: {
    //           enter(s) {
    //             // ___CSS_LOADER_EXPORT___
    //             const res = jsHandler(s.node.value, options)
    //             if (res.code) {
    //               s.node.value = res.code
    //             }
    //           },
    //         },
    //       })
    //     }
    //   },
    // },
  })

  return {
    code: ms.toString(),
    get map() {
      return ms.generateMap()
    },
  }
}
