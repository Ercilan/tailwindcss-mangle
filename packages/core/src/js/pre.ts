import babel from '@babel/core'
import { declare } from '@babel/helper-plugin-utils'
import MagicString from 'magic-string'
import { splitCode } from '@tailwindcss-mangle/shared'
import { sort } from 'fast-sort'
import { jsStringEscape } from '@ast-core/escape'
import { getStringLiteralCalleeName, getTemplateElementCalleeName } from './utils'
import type { Context } from '@/ctx'

interface Options {
  replaceMap: Map<string, string>
  magicString: MagicString
  id: string
  ctx: Context
}

type HandleValueOptions = {
  raw: string
  path: babel.NodePath<babel.types.StringLiteral | babel.types.TemplateElement>
  offset: number
  escape: boolean
  preserve: boolean
} & Options

export function handleValue(options: HandleValueOptions) {
  const { ctx, id, path, magicString, raw, replaceMap, offset = 0, escape = false, preserve = false } = options
  const node = path.node
  let value = raw
  const arr = sort(splitCode(value)).desc((x) => x.length)

  for (const str of arr) {
    if (replaceMap.has(str)) {
      ctx.addToUsedBy(str, id)
      if (preserve) {
        ctx.addPreserveClass(str)
      }
      // replace
      const v = replaceMap.get(str)
      if (v) {
        value = value.replaceAll(str, v)
      }
    }
  }
  if (preserve) {
    return
  }
  if (typeof node.start === 'number' && typeof node.end === 'number' && value) {
    const start = node.start + offset
    const end = node.end - offset
    if (start < end) {
      magicString.update(start, end, escape ? jsStringEscape(value) : value)
    }
  }
}

export const plugin = declare((api, options: Options) => {
  api.assertVersion(7)
  const { magicString, replaceMap, id, ctx } = options
  return {
    visitor: {
      StringLiteral: {
        enter(p) {
          const opts: HandleValueOptions = {
            ctx,
            id,
            magicString,
            path: p,
            raw: p.node.value,
            replaceMap,
            offset: 1,
            escape: true,
            preserve: false
          }
          const calleeName = getStringLiteralCalleeName(p)
          if (calleeName && ctx.isPreserveFunction(calleeName)) {
            opts.preserve = true
          }
          handleValue(opts)
        }
      },
      TemplateElement: {
        enter(p) {
          const opts: HandleValueOptions = {
            ctx,
            id,
            magicString,
            path: p,
            raw: p.node.value.raw,
            replaceMap,
            offset: 0,
            escape: false,
            preserve: false
          }
          const calleeName = getTemplateElementCalleeName(p)
          if (calleeName && ctx.isPreserveFunction(calleeName)) {
            opts.preserve = true
          }
          handleValue(opts)
        }
      }
    }
  }
})

export function preProcessJs(options: { code: string | MagicString; replaceMap: Map<string, string>; id: string; ctx: Context }) {
  const { code, replaceMap, id, ctx } = options
  const magicString = typeof code === 'string' ? new MagicString(code) : code

  babel.transformSync(magicString.original, {
    presets: [
      // ['@babel/preset-react', {}],
      [
        require('@babel/preset-typescript'),
        {
          allExtensions: true,
          isTSX: true
        }
      ]
    ],
    plugins: [
      [
        plugin,
        {
          magicString,
          replaceMap,
          id,
          ctx
        }
      ]
    ],
    filename: id
  })
  return magicString.toString()
}
