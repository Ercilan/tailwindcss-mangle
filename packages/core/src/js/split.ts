export const validateFilterRE = /[\w\u00A0-\uFFFF-_:%-?]/

export function isValidSelector(selector = ''): selector is string {
  return validateFilterRE.test(selector)
}

export const splitCode = (
  code: string,
  options: {
    splitQuote?: boolean
  } = { splitQuote: true }
) => {
  const regex = options.splitQuote ? /[\s"]+/ : /[\s]+/
  return code.split(regex).filter(isValidSelector)
}
