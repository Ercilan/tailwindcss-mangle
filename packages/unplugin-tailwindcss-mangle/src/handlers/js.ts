import { parse, traverse, generate } from '../lib/babel'

import type { Node } from '@babel/types'
import type { TraverseOptions, IJsHandlerOptions } from '../types'
import { escapeStringRegexp } from '../utils'
import { splitCode } from './split'

export function jsHandler(rawSource: string, options: IJsHandlerOptions) {
  const ast = parse(rawSource)
  const set = options.set
  const clsGen = options.classGenerator
  // 这样搞会把原先所有的 children 含有相关的 也都转义了
  const topt: TraverseOptions<Node> = {
    StringLiteral: {
      enter(p) {
        const n = p.node
        const arr = splitCode(n.value) // .split(/\s/).filter((x) => x)
        let rawStr = n.value
        for (let i = 0; i < arr.length; i++) {
          const v = arr[i]
          if (set.has(v)) {
            let ignoreFlag = false
            if (Array.isArray(n.leadingComments)) {
              ignoreFlag = n.leadingComments.findIndex((x) => x.value.includes('tw-mangle') && x.value.includes('ignore')) > -1
            }

            if (!ignoreFlag) {
              rawStr = rawStr.replace(new RegExp(escapeStringRegexp(v), 'g'), clsGen.generateClassName(v).name)
            }
          }
        }
        n.value = rawStr
      }
    },
    noScope: true
  }

  traverse(ast, topt)

  return generate(ast)
}
