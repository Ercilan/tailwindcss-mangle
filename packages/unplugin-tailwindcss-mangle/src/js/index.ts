import _generate from '@babel/generator'
import { parse } from '@babel/parser'
import _traverse from '@babel/traverse'

import type { Node, StringLiteral, TemplateElement } from '@babel/types'
import type { TraverseOptions, IHandlerOptions } from '../types'
import { escapeStringRegexp } from '../utils'
import { splitCode } from './split'

function getDefaultExportFromNamespaceIfPresent(n: any) {
  return n && Object.prototype.hasOwnProperty.call(n, 'default') ? n.default : n
}
const generate = getDefaultExportFromNamespaceIfPresent(_generate) as typeof _generate
const traverse = getDefaultExportFromNamespaceIfPresent(_traverse) as typeof _traverse

export function makeRegex(str: string) {
  return new RegExp('(?<=^|[\\s"])' + escapeStringRegexp(str), 'g')
}

export function handleValue(str: string, node: StringLiteral | TemplateElement, options: IHandlerOptions) {
  const set = options.runtimeSet
  const clsGen = options.classGenerator
  const arr = splitCode(str)
  let rawStr = str
  for (let i = 0; i < arr.length; i++) {
    const v = arr[i]
    if (set.has(v)) {
      let ignoreFlag = false
      if (Array.isArray(node.leadingComments)) {
        ignoreFlag = node.leadingComments.findIndex((x) => x.value.includes('tw-mangle') && x.value.includes('ignore')) > -1
      }

      if (!ignoreFlag) {
        rawStr = rawStr.replace(makeRegex(v), clsGen.generateClassName(v).name)
      }
    }
  }
  return rawStr
}

export function jsHandler(rawSource: string, options: IHandlerOptions) {
  const ast = parse(rawSource)

  const topt: TraverseOptions<Node> = {
    StringLiteral: {
      enter(p) {
        const n = p.node
        n.value = handleValue(n.value, n, options)
      }
    },
    TemplateElement: {
      enter(p) {
        const n = p.node
        n.value.raw = handleValue(n.value.raw, n, options)
      }
    },
    noScope: true
  }

  traverse(ast, topt)

  return generate(ast)
}
