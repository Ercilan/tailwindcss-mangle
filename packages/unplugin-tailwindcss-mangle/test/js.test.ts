import { jsHandler } from '../src/js/index'
import ClassGenerator from '../src/classGenerator'
describe('js handler', () => {
  it('common StringLiteral', () => {
    const classGenerator = new ClassGenerator()
    const runtimeSet = new Set<string>()
    runtimeSet.add('dark:bg-zinc-800/30')
    runtimeSet.add('lg:dark:bg-zinc-800/30')
    // eslint-disable-next-line no-template-curly-in-string
    const testCase = 'element.innerHTML = \'<div class="dark:bg-zinc-800/30 lg:dark:bg-zinc-800/30">count is counter</div>\''
    const code = jsHandler(testCase, {
      classGenerator,
      runtimeSet
    }).code
    expect(code).toMatchSnapshot()
  })

  it('common TemplateElement', () => {
    const classGenerator = new ClassGenerator()
    const runtimeSet = new Set<string>()
    runtimeSet.add('dark:bg-zinc-800/30')
    runtimeSet.add('lg:dark:bg-zinc-800/30')
    // eslint-disable-next-line no-template-curly-in-string
    const testCase = 'const counter = 0;element.innerHTML = `<div class="dark:bg-zinc-800/30 lg:dark:bg-zinc-800/30">count is ${counter}</div>`'
    const code = jsHandler(testCase, {
      classGenerator,
      runtimeSet
    }).code
    expect(code).toMatchSnapshot()
  })

  it('z-10 not transform', () => {
    const classGenerator = new ClassGenerator()
    const runtimeSet = new Set<string>()
    const arr = 'z-10 w-full max-w-5xl items-center justify-between font-mono text-sm lg:flex'.split(' ')
    for (let i = 0; i < arr.length; i++) {
      const element = arr[i]
      runtimeSet.add(element)
    }

    const testCase = `{ className: "z-10 w-full max-w-5xl items-center justify-between font-mono text-sm lg:flex" }`
    const code = jsHandler(testCase, {
      classGenerator,
      runtimeSet
    }).code
    expect(code).toMatchSnapshot()
  })
})
