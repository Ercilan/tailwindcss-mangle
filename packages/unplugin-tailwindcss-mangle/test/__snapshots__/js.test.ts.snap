// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`js handler common StringLiteral 1`] = `"element.innerHTML = "<div class=\\"tw-a tw-b\\">count is counter</div>";"`;

exports[`js handler common TemplateElement 1`] = `
"const counter = 0;
element.innerHTML = \`<div class="tw-a tw-b">count is \${counter}</div>\`;"
`;

exports[`js handler nextjs server side mangle 1`] = `
""use strict";

(() => {
  var exports = {};
  exports.id = 405;
  exports.ids = [405];
  exports.modules = {
    /***/957: /***/(__unused_webpack_module, __webpack_exports__, __webpack_require__) => {
      __webpack_require__.r(__webpack_exports__);
      /* harmony export */
      __webpack_require__.d(__webpack_exports__, {
        /* harmony export */"default": () => /* binding */Home
        /* harmony export */
      });
      /* harmony import */
      var react_jsx_runtime__WEBPACK_IMPORTED_MODULE_0__ = __webpack_require__(997);
      /* harmony import */
      var react_jsx_runtime__WEBPACK_IMPORTED_MODULE_0___default = /*#__PURE__*/__webpack_require__.n(react_jsx_runtime__WEBPACK_IMPORTED_MODULE_0__);
      /* harmony import */
      var next_font_google_target_css_path_pages_index_tsx_import_Inter_arguments_subsets_latin_variableName_inter___WEBPACK_IMPORTED_MODULE_2__ = __webpack_require__(941);
      /* harmony import */
      var next_font_google_target_css_path_pages_index_tsx_import_Inter_arguments_subsets_latin_variableName_inter___WEBPACK_IMPORTED_MODULE_2___default = /*#__PURE__*/__webpack_require__.n(next_font_google_target_css_path_pages_index_tsx_import_Inter_arguments_subsets_latin_variableName_inter___WEBPACK_IMPORTED_MODULE_2__);
      /* harmony import */
      var next_image__WEBPACK_IMPORTED_MODULE_1__ = __webpack_require__(395);
      /* harmony import */
      var next_image__WEBPACK_IMPORTED_MODULE_1___default = /*#__PURE__*/__webpack_require__.n(next_image__WEBPACK_IMPORTED_MODULE_1__);
      function Home() {
        return /*#__PURE__*/(0, react_jsx_runtime__WEBPACK_IMPORTED_MODULE_0__.jsxs)("main", {
          className: "flex tw-a tw-b tw-c tw-d tw-e",
          children: [/*#__PURE__*/(0, react_jsx_runtime__WEBPACK_IMPORTED_MODULE_0__.jsxs)("div", {
            className: "tw-f tw-g tw-h tw-c tw-d tw-i tw-j tw-k",
            children: [/*#__PURE__*/(0, react_jsx_runtime__WEBPACK_IMPORTED_MODULE_0__.jsxs)("p", {
              className: "fixed tw-l tw-m flex tw-g tw-n tw-o tw-p tw-q tw-r tw-s tw-t tw-u tw-v tw-w tw-x tw-y tw-z  tw-aa tw-ba tw-ca tw-da tw-ea",
              children: ["Get started by editing\\xa0", /*#__PURE__*/react_jsx_runtime__WEBPACK_IMPORTED_MODULE_0__.jsx("code", {
                className: "tw-i tw-fa",
                children: "pages/index.tsx"
              })]
            }), /*#__PURE__*/react_jsx_runtime__WEBPACK_IMPORTED_MODULE_0__.jsx("div", {
              className: "fixed tw-ga tw-l flex tw-ha tw-g tw-ia tw-n tw-ja tw-ka tw-la tw-ma tw-na tw-y tw-oa tw-z tw-pa",
              children: /*#__PURE__*/(0, react_jsx_runtime__WEBPACK_IMPORTED_MODULE_0__.jsxs)("a", {
                className: "tw-qa flex tw-ra tw-sa tw-ta tw-ua tw-va",
                href: "https://vercel.com?utm_source=create-next-app&utm_medium=default-template-tw&utm_campaign=create-next-app",
                target: "_blank",
                rel: "noopener noreferrer",
                children: ["By ", /*#__PURE__*/react_jsx_runtime__WEBPACK_IMPORTED_MODULE_0__.jsx(next_image__WEBPACK_IMPORTED_MODULE_1___default(), {
                  src: "/vercel.svg",
                  alt: "Vercel Logo",
                  className: "tw-wa",
                  width: 100,
                  height: 24,
                  priority: true
                })]
              })
            })]
          }), /*#__PURE__*/react_jsx_runtime__WEBPACK_IMPORTED_MODULE_0__.jsx("div", {
            className: "relative flex tw-ra tw-xa tw-ya tw-za tw-ab tw-bb tw-cb tw-db tw-eb tw-fb tw-gb tw-hb tw-ib tw-jb tw-kb tw-lb tw-mb tw-nb tw-ob tw-pb tw-qb tw-rb tw-sb tw-tb tw-ub tw-vb tw-wb",
            children: /*#__PURE__*/react_jsx_runtime__WEBPACK_IMPORTED_MODULE_0__.jsx(next_image__WEBPACK_IMPORTED_MODULE_1___default(), {
              className: "relative tw-xb tw-wa",
              src: "/next.svg",
              alt: "Next.js Logo",
              width: 180,
              height: 37,
              priority: true
            })
          }), /*#__PURE__*/(0, react_jsx_runtime__WEBPACK_IMPORTED_MODULE_0__.jsxs)("div", {
            className: "tw-yb grid tw-zb tw-ac tw-bc tw-cc",
            children: [/*#__PURE__*/(0, react_jsx_runtime__WEBPACK_IMPORTED_MODULE_0__.jsxs)("a", {
              href: "https://nextjs.org/docs?utm_source=create-next-app&utm_medium=default-template-tw&utm_campaign=create-next-app",
              className: "group tw-dc border tw-ec tw-fc tw-gc tw-hc tw-ic tw-jc tw-kc tw-lc",
              target: "_blank",
              rel: "noopener noreferrer",
              children: [/*#__PURE__*/(0, react_jsx_runtime__WEBPACK_IMPORTED_MODULE_0__.jsxs)("h2", {
                className: \`\${next_font_google_target_css_path_pages_index_tsx_import_Inter_arguments_subsets_latin_variableName_inter___WEBPACK_IMPORTED_MODULE_2___default().className} tw-mc tw-nc tw-oc\`,
                children: ["Docs ", /*#__PURE__*/react_jsx_runtime__WEBPACK_IMPORTED_MODULE_0__.jsx("span", {
                  className: "tw-pc tw-qc tw-rc tw-sc",
                  children: "->"
                })]
              }), /*#__PURE__*/react_jsx_runtime__WEBPACK_IMPORTED_MODULE_0__.jsx("p", {
                className: \`\${next_font_google_target_css_path_pages_index_tsx_import_Inter_arguments_subsets_latin_variableName_inter___WEBPACK_IMPORTED_MODULE_2___default().className} tw-tc tw-uc tw-j tw-vc\`,
                children: "Find in-depth information about Next.js features and API."
              })]
            }), /*#__PURE__*/(0, react_jsx_runtime__WEBPACK_IMPORTED_MODULE_0__.jsxs)("a", {
              href: "https://nextjs.org/learn?utm_source=create-next-app&utm_medium=default-template-tw&utm_campaign=create-next-app",
              className: "group tw-dc border tw-ec tw-fc tw-gc tw-hc tw-ic tw-jc tw-kc tw-lc",
              target: "_blank",
              rel: "noopener noreferrer",
              children: [/*#__PURE__*/(0, react_jsx_runtime__WEBPACK_IMPORTED_MODULE_0__.jsxs)("h2", {
                className: \`\${next_font_google_target_css_path_pages_index_tsx_import_Inter_arguments_subsets_latin_variableName_inter___WEBPACK_IMPORTED_MODULE_2___default().className} tw-mc tw-nc tw-oc\`,
                children: ["Learn ", /*#__PURE__*/react_jsx_runtime__WEBPACK_IMPORTED_MODULE_0__.jsx("span", {
                  className: "tw-pc tw-qc tw-rc tw-sc",
                  children: "->"
                })]
              }), /*#__PURE__*/react_jsx_runtime__WEBPACK_IMPORTED_MODULE_0__.jsx("p", {
                className: \`\${next_font_google_target_css_path_pages_index_tsx_import_Inter_arguments_subsets_latin_variableName_inter___WEBPACK_IMPORTED_MODULE_2___default().className} tw-tc tw-uc tw-j tw-vc\`,
                children: "Learn about Next.js in an interactive course with\\xa0quizzes!"
              })]
            }), /*#__PURE__*/(0, react_jsx_runtime__WEBPACK_IMPORTED_MODULE_0__.jsxs)("a", {
              href: "https://vercel.com/templates?framework=next.js&utm_source=create-next-app&utm_medium=default-template-tw&utm_campaign=create-next-app",
              className: "group tw-dc border tw-ec tw-fc tw-gc tw-hc tw-ic tw-jc tw-kc tw-lc",
              target: "_blank",
              rel: "noopener noreferrer",
              children: [/*#__PURE__*/(0, react_jsx_runtime__WEBPACK_IMPORTED_MODULE_0__.jsxs)("h2", {
                className: \`\${next_font_google_target_css_path_pages_index_tsx_import_Inter_arguments_subsets_latin_variableName_inter___WEBPACK_IMPORTED_MODULE_2___default().className} tw-mc tw-nc tw-oc\`,
                children: ["Templates ", /*#__PURE__*/react_jsx_runtime__WEBPACK_IMPORTED_MODULE_0__.jsx("span", {
                  className: "tw-pc tw-qc tw-rc tw-sc",
                  children: "->"
                })]
              }), /*#__PURE__*/react_jsx_runtime__WEBPACK_IMPORTED_MODULE_0__.jsx("p", {
                className: \`\${next_font_google_target_css_path_pages_index_tsx_import_Inter_arguments_subsets_latin_variableName_inter___WEBPACK_IMPORTED_MODULE_2___default().className} tw-tc tw-uc tw-j tw-vc\`,
                children: "Discover and deploy boilerplate example Next.js\\xa0projects."
              })]
            }), /*#__PURE__*/(0, react_jsx_runtime__WEBPACK_IMPORTED_MODULE_0__.jsxs)("a", {
              href: "https://vercel.com/new?utm_source=create-next-app&utm_medium=default-template-tw&utm_campaign=create-next-app",
              className: "group tw-dc border tw-ec tw-fc tw-gc tw-hc tw-ic tw-jc tw-kc tw-lc",
              target: "_blank",
              rel: "noopener noreferrer",
              children: [/*#__PURE__*/(0, react_jsx_runtime__WEBPACK_IMPORTED_MODULE_0__.jsxs)("h2", {
                className: \`\${next_font_google_target_css_path_pages_index_tsx_import_Inter_arguments_subsets_latin_variableName_inter___WEBPACK_IMPORTED_MODULE_2___default().className} tw-mc tw-nc tw-oc\`,
                children: ["Deploy ", /*#__PURE__*/react_jsx_runtime__WEBPACK_IMPORTED_MODULE_0__.jsx("span", {
                  className: "tw-pc tw-qc tw-rc tw-sc",
                  children: "->"
                })]
              }), /*#__PURE__*/react_jsx_runtime__WEBPACK_IMPORTED_MODULE_0__.jsx("p", {
                className: \`\${next_font_google_target_css_path_pages_index_tsx_import_Inter_arguments_subsets_latin_variableName_inter___WEBPACK_IMPORTED_MODULE_2___default().className} tw-tc tw-uc tw-j tw-vc\`,
                children: "Instantly deploy your Next.js site to a shareable URL with Vercel."
              })]
            })]
          })]
        });
      }

      /***/
    },

    /***/918: /***/module => {
      module.exports = require("next/dist/shared/lib/amp-context.js");

      /***/
    },

    /***/732: /***/module => {
      module.exports = require("next/dist/shared/lib/amp-mode.js");

      /***/
    },

    /***/796: /***/module => {
      module.exports = require("next/dist/shared/lib/head-manager-context.js");

      /***/
    },

    /***/486: /***/module => {
      module.exports = require("next/dist/shared/lib/image-blur-svg.js");

      /***/
    },

    /***/744: /***/module => {
      module.exports = require("next/dist/shared/lib/image-config-context.js");

      /***/
    },

    /***/843: /***/module => {
      module.exports = require("next/dist/shared/lib/image-config.js");

      /***/
    },

    /***/552: /***/module => {
      module.exports = require("next/dist/shared/lib/image-loader");

      /***/
    },

    /***/470: /***/module => {
      module.exports = require("next/dist/shared/lib/side-effect.js");

      /***/
    },

    /***/618: /***/module => {
      module.exports = require("next/dist/shared/lib/utils/warn-once.js");

      /***/
    },

    /***/689: /***/module => {
      module.exports = require("react");

      /***/
    },

    /***/997: /***/module => {
      module.exports = require("react/jsx-runtime");

      /***/
    }
  };

  ;

  // load runtime
  var __webpack_require__ = require("../webpack-runtime.js");
  __webpack_require__.C(exports);
  var __webpack_exec__ = moduleId => __webpack_require__(__webpack_require__.s = moduleId);
  var __webpack_exports__ = __webpack_require__.X(0, [463, 262], () => __webpack_exec__(957));
  module.exports = __webpack_exports__;
})();"
`;

exports[`js handler z-10 not transform 1`] = `
"{
  className: "tw-a tw-b tw-c tw-d tw-e tw-f tw-g tw-h";
}"
`;
